src/audio/component.c

```diff
@@ -328,162 +328,273 @@ int comp_verify_params(struct comp_dev *dev, uint32_t flag,
+		return -EBUSY;
+	}
+
 	/* in case when the current package is the first, we should allocate
 	 * memory for whole model data
 	 */
 	if (!cdata->msg_index) {
-		ret = comp_alloc_model_data(dev, model, cdata->data->size);
 		/* in case when required model size is equal to zero we do not
 		 * allocate memory and should just return 0
 		 */
-		if (ret < 0 || !cdata->data->size)
-			return ret;
+		if (!cdata->data->size)
+			return 0;
+
+		blob_handler->data_new = rballoc(0, SOF_MEM_CAPS_RAM,
+						 cdata->data->size);
+		if (!blob_handler->data_new) {
+			comp_err(blob_handler->dev, "comp_data_blob_set_cmd(): blob_handler->data_new allocation failed.");
+			return -ENOMEM;
+		}
+
+		blob_handler->data_size = cdata->data->size;
+		blob_handler->data_ready = false;
+		blob_handler->data_pos = 0;
 	}
 
 	/* return an error in case when we do not have allocated memory for
 	 * model data
 	 */
-	if (!model->data) {
-		comp_err(dev, "comp_set_model(): buffer not allocated");
+	if (!blob_handler->data_new) {
+		comp_err(blob_handler->dev, "comp_data_blob_set_cmd(): buffer not allocated");
 		return -ENOMEM;
 	}
 
-	if (!cdata->elems_remaining) {
-		/* when we receive the last package and do not fill the whole
-		 * allocated buffer, we return an error
-		 */
-		if (cdata->num_elems + model->data_pos < model->data_size) {
-			comp_err(dev, "comp_set_model(): not enough data to fill the buffer");
-			// TODO: handle this situation
-
-			return -EINVAL;
-		}
+	ret = memcpy_s((char *)blob_handler->data_new + blob_handler->data_pos,
+		       blob_handler->data_size - blob_handler->data_pos,
+		       cdata->data->data, cdata->num_elems);
+	assert(!ret);
 
-		/* the whole data were received properly */
-		done = true;
-		comp_dbg(dev, "comp_set_model(): final package received");
-	}
+	blob_handler->data_pos += cdata->num_elems;
 
-	/* return an error in case when received data exceed allocated
-	 * memory
-	 */
-	if (cdata->num_elems > model->data_size - model->data_pos) {
-		comp_err(dev, "comp_set_model(): too much data");
-		return -EINVAL;
-	}
+	if (!cdata->elems_remaining) {
+		comp_dbg(blob_handler->dev, "comp_data_blob_set_cmd(): final package received");
 
-	ret = memcpy_s((char *)model->data + model->data_pos,
-		       model->data_size - model->data_pos,
-		       cdata->data->data, cdata->num_elems);
-	assert(!ret);
+		/* The new configuration is OK to be applied */
+		blob_handler->data_ready = true;
 
-	/* update data_pos variable with received number of elements (num_elem)
-	 */
-	model->data_pos += cdata->num_elems;
+		/* If component state is READY we can omit old
+		 * configuration immediately. When in playback/capture
+		 * the new configuration presence is checked in copy().
+		 */
+		if (blob_handler->dev->state ==  COMP_STATE_READY) {
+			rfree(blob_handler->data);
+			blob_handler->data = NULL;
+		}
 
-	/* Update crc value when done */
-	if (done) {
-		model->crc = crc32(0, model->data, model->data_size);
```

Comment:
What's happening to the CRC ?

Reply:
@lgirdwood 
CRC is done directly in component:
``` c
static int test_keyword_prepare(struct comp_dev *dev)
{
	struct comp_data *cd = comp_get_drvdata(dev);
	uint16_t valid_bits = cd->sample_valid_bytes * 8;
	uint16_t sample_width = cd->config.sample_width;

	comp_info(dev, "test_keyword_prepare()");

	if (valid_bits != sample_width) {
		/* Default threshold value has to be changed
		 * according to host new format.
		 */
		cd->config.activation_threshold =
			test_keyword_get_threshold(dev, valid_bits);
	}

	cd->data_blob = comp_get_data_blob(cd->model_handler,
					   &cd->data_blob_size);
	if (cd->data_blob)
		cd->data_blob_crc = crc32(0, cd->data_blob, cd->data_blob_size);

	return comp_set_state(dev, COMP_TRIGGER_PREPARE);
}
```
This generic methods returns only pointer to data blob with its size. If some component wants to make data CRC, it should do it itself.

Reply:
ok, can
```c
if (cd->data_blob)
	cd->data_blob_crc = crc32(0, cd->data_blob, cd->data_blob_size);
```
be part of
```c
cd->data_blob = comp_get_data_blob(cd->model_handler,
					   &cd->data_blob_size);
```
This way this CRC is always done by comp core as part of the get().

Reply:
@lgirdwood 
Ok, I will update it.

Reply:
@lgirdwood 
Updated